### ISSUE TITLE
Software Cursor trouble in Java awt

### ISSUE THREAD

**Issue by ChristophHaag** [Gnurou/xserver#1]
The same problem as described here https://bugs.launchpad.net/unity-mir/+bug/1250613.

The test program that is mentioned there exhibits the same problem:

```
import java.awt.dnd.DragSource;
public class MirBug {
  public static void main(String[] args) {
    DragSource.getDefaultDragSource(); // Eventually call sun.awt.X11.XToolkit.lazilyLoadDesktopProperty
  }
}
```

Save as MirBug.java, run `javac MirBug.java; java -cp . MirBug` and it will throw a runtime exception.

Can you do anything about that in the X server?

**Comment by Gnurou** [comment]
This is not the official xserver repo, and I am not sure which problem you are referring to?

**Comment by ChristophHaag** [comment]
Running that test program just exits normally without doing anything on my x86 laptop with intel GPU, as it should.

On my Tegra K1 Chromebook with nouveau it does this:

```
$ javac MirBug.java; java -cp . MirBug
java.lang.RuntimeException: cannot load system cursor: CopyDrop.32x32
    at sun.awt.X11.XToolkit.lazilyLoadDesktopProperty(XToolkit.java:1541)
    at java.awt.Toolkit.getDesktopProperty(Toolkit.java:1803)
    at java.awt.dnd.DragSource.load(DragSource.java:131)
    at java.awt.dnd.DragSource.<clinit>(DragSource.java:148)
    at MirBug.main(MirBug.java:5)
Caused by: java.awt.AWTException: Exception: class java.lang.IllegalArgumentException Width (0) and height (0) must be non-zero occurred while creating cursor CopyDrop.32x32
    at java.awt.Cursor.getSystemCustomCursor(Cursor.java:363)
    at sun.awt.X11.XToolkit.lazilyLoadDesktopProperty(XToolkit.java:1539)
    ... 4 more
Exception in thread "main" java.lang.ExceptionInInitializerError
    at MirBug.main(MirBug.java:5)
Caused by: java.lang.RuntimeException: failed to load system cursor: DnD.Cursor.CopyDrop : cannot load system cursor: CopyDrop.32x32
    at java.awt.dnd.DragSource.load(DragSource.java:135)
    at java.awt.dnd.DragSource.<clinit>(DragSource.java:148)
    ... 1 more
```

From the comments on the bug report it seems it has something to do with software cursor handling in X11 and I think that it is caused by the modesetting ddx running on tegradrm somehow. Maybe it is even a java bug, but the thing is that because of that many java AWT applications can only be used inside a vnc server on at least the Tegra K1.

I tried reading a bit through the openjdk source but it's actually not that easy to read...

As per comment 2 in that bug report, maybe it'd be the easiest workaround to just make a library with a hacked implementation of XQueryBestCursor and LD_PRELOAD that.

**Comment by ChristophHaag** [comment]
First time I've done something like this:

CursorFix.c

``` c
// gcc CursorFix.c -lX11 -fPIC -shared -Wl,-soname,libX11.so -ldl -o libX11.so

// Run with LD_PRELOAD=/wherever/libX11.so java ...

#define _GNU_SOURCE
#include <dlfcn.h>
#include <stdio.h>
#include "X11/Xlibint.h"

typedef int (*orig_XQueryBestCursor_f_type)(
    register Display *dpy,
    Drawable drawable,
    unsigned int width,
    unsigned int height,
    unsigned int *ret_width,
    unsigned int *ret_height);

Status XQueryBestCursor(
    register Display *dpy,
    Drawable drawable,
    unsigned int width,
    unsigned int height,
    unsigned int *ret_width,
    unsigned int *ret_height)
{
    orig_XQueryBestCursor_f_type orig_XQueryBestCursor;
    orig_XQueryBestCursor = (orig_XQueryBestCursor_f_type)dlsym(RTLD_NEXT,"XQueryBestCursor");
    Status retval = orig_XQueryBestCursor(
    dpy,
    drawable,
    width,
    height,
    ret_width,
    ret_height);

    printf("Fixing up values...\n");

    *ret_width = 32;
    *ret_height = 32;

    return retval;
}
```

And Intellij seems to work so far.

**Comment by ChristophHaag** [comment]
So I guess this is a bug in the modesetting driver?

I googled a bit and found https://bugs.launchpad.net/ubuntu/+source/xorg-server/+bug/1246872

It says that this is returning what xf86_cursors_init is setting.

~~Looking at modesetting's ScreenInit function there is only this: https://github.com/Gnurou/xserver/blob/b381eab072c8eb9738397b3a2639e09213b3cd70/hw/xfree86/drivers/modesetting/driver.c#L1198-L1203~~

~~So the cursor size is never set when there's no hardware cursor? tegradrm on Tegra K1 is using a software cursor, right? I'll test later whether it's enough to simply set the size in the software cursor size too.~~

Edit: This was wrong. `ms->drmmode.sw_cursor` is 0, so the `xf86_cursors_init` function IS executed. I also looked at the arguments and it's 64x64. I tried hacking in `ms->cursor_height = 32;` and `ms->cursor_width = 32;` just before that call to confirm that it's not a weird mismatch issue somewhere, but the result is the same.
Looks like these values are getting lost somewhere in the modesetting driver or in xlib, so I suppose it really is an upstream issue in X. Still one that happens on my Tegra K1 chromebook so not entirely unrelated. :)
